VLA issue with 'MG_ENABLE_TCPIP'

Compiling with MSVC and `-DMG_ENABLE_TCPIP=1`, gives me this error in `src/net_builtin.c`:
```c
net_builtin.c(288): error C2057: expected constant expression
net_builtin.c(288): error C2466: cannot allocate an array of constant size 0
net_builtin.c(288): error C2133: 'opts': unknown size
```

But why not simply define `opts_maxlen` as a pre-processor constant:
```diff
--- a/src/net_builtin.c 2025-04-09 11:55:26
+++ b/src/net_builtin.c 2025-04-09 12:53:12
@@ -284,8 +284,8 @@
                              (ifp->enable_req_sntp ? 1 : 0));
   size_t len = strlen(ifp->dhcp_name);
   size_t olen = 21 + len + extra + 2 + 1;   // Total length of options
-  uint8_t opts_maxlen = 21 + sizeof(ifp->dhcp_name) + 2 + 2 + 1;
-  uint8_t opts[opts_maxlen]; // Allocate options (max size possible)
+  #define OPTS_MAXLEN (21 + sizeof(ifp->dhcp_name) + 2 + 2 + 1)
+  uint8_t opts[OPTS_MAXLEN]; // Allocate options (max size possible)
   uint8_t *p = opts;
   assert(olen <= sizeof(opts));
   memset(opts, 0, sizeof(opts));
```

Besides, I had this issue with `src/arch_win32.h` messing up with the  `-DMG_ENABLE_TCPIP` code:
```c
.\src/net_builtin.c(924,52): error: too few arguments provided to function-like macro invocation
  924 |     bool up = ifp->driver->poll(ifp, expired_1000ms);
      |                                                    ^
.\src\arch_win32.h(85,9): note: macro 'poll' defined here
   85 | #define poll(a, b, c) WSAPoll((a), (b), (c))
      |         ^
1 error generated.
```

I fail to understand the reason for this. But patched it like so:
```diff
--- a/src/arch_win32.h 2024-11-14 14:53:41
+++ b/src/arch_win32.h 2025-04-09 13:02:15
@@ -77,12 +77,16 @@
 #define MG_SOCKET_TYPE SOCKET
 typedef unsigned long nfds_t;
 #if defined(_MSC_VER)
+#if (MG_ENABLE_TCPIP == 0)
 #pragma comment(lib, "ws2_32.lib")
+#endif
 #ifndef alloca
 #define alloca(a) _alloca(a)
 #endif
 #endif
+#if (MG_ENABLE_TCPIP == 0)
 #define poll(a, b, c) WSAPoll((a), (b), (c))
+#endif
 #define closesocket(x) closesocket(x)
```

(what has `ws2_32.lib` to do with built-in networking?)

So after these changes, this sample works fine on Windows:
```
pcap-driver.exe -v 5 -i "\Device\NPF_{6568F487-0016-453C-8C29-5C28DEDF27A0}"
430cfa 2 main.c:267:main                Opened interface \Device\NPF_{6568F487-0016-453C-8C29-5C28DEDF27A0}
430cfa 2 main.c:293:main                Init done, starting main loop
430cfa 3 net.c:198:mg_listen            1 0 http://0.0.0.0:8000
430cfa 3 net.c:198:mg_listen            2 0 https://0.0.0.0:8443
430cfa 1 net_builtin.c:214:onstatechang Link up
430d0a 3 net_builtin.c:324:tx_dhcp_disc DHCP discover sent. Our MAC: 02:00:01:02:03:77
4310e2 3 net_builtin.c:324:tx_dhcp_disc DHCP discover sent. Our MAC: 02:00:01:02:03:77
4314ca 3 net_builtin.c:324:tx_dhcp_disc DHCP discover sent. Our MAC: 02:00:01:02:03:77
4318b2 3 net_builtin.c:324:tx_dhcp_disc DHCP discover sent. Our MAC: 02:00:01:02:03:77
4319bc 3 net_builtin.c:303:tx_dhcp_requ DHCP req sent
431a0a 2 net_builtin.c:442:rx_dhcp_clie Lease: 86400 sec (90797)
431a0a 1 net_builtin.c:211:onstatechang Got IP
431a29 2 net_builtin.c:207:onstatechang READY, IP: 10.0.0.147
431a29 2 net_builtin.c:208:onstatechang        GW: 10.0.0.1
431a29 2 net_builtin.c:209:onstatechang       MAC: 02:00:01:02:03:77
...
```

--- a/mongoose.h
+++ b/mongoose.h
@@ -424,6 +424,18 @@ static inline int mg_mkdir(const char *path, mode_t mode) {
 
 #if MG_ARCH == MG_ARCH_WIN32
 
+// Avoid name clashing; (macro expansion producing 'defined' has undefined
+// behaviour). See config.h for user options
+#ifndef MG_ENABLE_WINSOCK
+#if (!defined(MG_ENABLE_TCPIP) || !MG_ENABLE_TCPIP) && \
+    (!defined(MG_ENABLE_LWIP) || !MG_ENABLE_LWIP) &&   \
+    (!defined(MG_ENABLE_FREERTOS_TCP) || !MG_ENABLE_FREERTOS_TCP)
+#define MG_ENABLE_WINSOCK 1
+#else
+#define MG_ENABLE_WINSOCK 0
+#endif
+#endif
+
 #ifndef _CRT_RAND_S
 #define _CRT_RAND_S
 #endif
@@ -469,17 +481,19 @@ typedef enum { false = 0, true = 1 } bool;
 #else
 #include <stdbool.h>
 #include <stdint.h>
+#if MG_ENABLE_WINSOCK
 #include <ws2tcpip.h>
+#endif
 #endif
 
 #include <process.h>
 #include <winerror.h>
-#include <winsock2.h>
 
 // For mg_random()
 #if defined(_MSC_VER) && _MSC_VER < 1700
 #ifndef _WIN32_WINNT
 #define _WIN32_WINNT 0x400  // Let vc98 pick up wincrypt.h
+#include <winsock2.h>       // and fix missing macros
 #endif
 #include <wincrypt.h>
 #pragma comment(lib, "advapi32.lib")
@@ -495,24 +509,30 @@ typedef enum { false = 0, true = 1 } bool;
 #endif
 #endif
 
-#define MG_INVALID_SOCKET INVALID_SOCKET
-#define MG_SOCKET_TYPE SOCKET
 typedef unsigned long nfds_t;
 #if defined(_MSC_VER)
+#if MG_ENABLE_WINSOCK
 #pragma comment(lib, "ws2_32.lib")
+#endif
 #ifndef alloca
 #define alloca(a) _alloca(a)
 #endif
 #endif
-#define poll(a, b, c) WSAPoll((a), (b), (c))
-#define closesocket(x) closesocket(x)
 
-typedef int socklen_t;
 #define MG_DIRSEP '\\'
 
 #ifndef MG_PATH_MAX
 #define MG_PATH_MAX FILENAME_MAX
 #endif
+
+#if MG_ENABLE_WINSOCK
+#include <winsock2.h>
+
+#define MG_INVALID_SOCKET INVALID_SOCKET
+#define MG_SOCKET_TYPE SOCKET
+#define poll(a, b, c) WSAPoll((a), (b), (c))
+#define closesocket(x) closesocket(x)
+typedef int socklen_t;
 
 #ifndef SO_EXCLUSIVEADDRUSE
 #define SO_EXCLUSIVEADDRUSE ((int) (~SO_REUSEADDR))
@@ -528,13 +548,15 @@ typedef int socklen_t;
 #define MG_SOCK_RESET(errcode) \
   (((errcode) < 0) && (WSAGetLastError() == WSAECONNRESET))
 
+#endif  // MG_ENABLE_WINSOCK
+
 #define realpath(a, b) _fullpath((b), (a), MG_PATH_MAX)
-#define sleep(x) Sleep((x) * 1000)
+#define sleep(x) Sleep((x) *1000)
 #define mkdir(a, b) _mkdir(a)
 #define timegm(x) _mkgmtime(x)
 
 #ifndef S_ISDIR
-#define S_ISDIR(x) (((x) & _S_IFMT) == _S_IFDIR)
+#define S_ISDIR(x) (((x) &_S_IFMT) == _S_IFDIR)
 #endif
 
 #ifndef MG_ENABLE_DIRLIST
@@ -774,7 +796,8 @@ struct timeval {
 #define MG_ENABLE_MD5 1
 #endif
 
-// Set MG_ENABLE_WINSOCK=0 for Win32 builds with external IP stack (like LWIP)
+// Set MG_ENABLE_WINSOCK=0 for Win32 builds with other external IP stack not
+// mentioned in arch_win32.h
 #ifndef MG_ENABLE_WINSOCK
 #define MG_ENABLE_WINSOCK 1
 #endif

--- a/src/arch_win32.h
+++ b/src/arch_win32.h
@@ -2,6 +2,18 @@
 
 #if MG_ARCH == MG_ARCH_WIN32
 
+// Avoid name clashing; (macro expansion producing 'defined' has undefined
+// behaviour). See config.h for user options
+#ifndef MG_ENABLE_WINSOCK
+#if (!defined(MG_ENABLE_TCPIP) || !MG_ENABLE_TCPIP) && \
+    (!defined(MG_ENABLE_LWIP) || !MG_ENABLE_LWIP) &&   \
+    (!defined(MG_ENABLE_FREERTOS_TCP) || !MG_ENABLE_FREERTOS_TCP)
+#define MG_ENABLE_WINSOCK 1
+#else
+#define MG_ENABLE_WINSOCK 0
+#endif
+#endif
+
 #ifndef _CRT_RAND_S
 #define _CRT_RAND_S
 #endif
@@ -47,17 +59,19 @@ typedef enum { false = 0, true = 1 } bool;
 #else
 #include <stdbool.h>
 #include <stdint.h>
+#if MG_ENABLE_WINSOCK
 #include <ws2tcpip.h>
+#endif
 #endif
 
 #include <process.h>
 #include <winerror.h>
-#include <winsock2.h>
 
 // For mg_random()
 #if defined(_MSC_VER) && _MSC_VER < 1700
 #ifndef _WIN32_WINNT
 #define _WIN32_WINNT 0x400  // Let vc98 pick up wincrypt.h
+#include <winsock2.h>       // and fix missing macros
 #endif
 #include <wincrypt.h>
 #pragma comment(lib, "advapi32.lib")
@@ -73,24 +87,30 @@ typedef enum { false = 0, true = 1 } bool;
 #endif
 #endif
 
-#define MG_INVALID_SOCKET INVALID_SOCKET
-#define MG_SOCKET_TYPE SOCKET
 typedef unsigned long nfds_t;
 #if defined(_MSC_VER)
+#if MG_ENABLE_WINSOCK
 #pragma comment(lib, "ws2_32.lib")
+#endif
 #ifndef alloca
 #define alloca(a) _alloca(a)
 #endif
 #endif
-#define poll(a, b, c) WSAPoll((a), (b), (c))
-#define closesocket(x) closesocket(x)
 
-typedef int socklen_t;
 #define MG_DIRSEP '\\'
 
 #ifndef MG_PATH_MAX
 #define MG_PATH_MAX FILENAME_MAX
 #endif
+
+#if MG_ENABLE_WINSOCK
+#include <winsock2.h>
+
+#define MG_INVALID_SOCKET INVALID_SOCKET
+#define MG_SOCKET_TYPE SOCKET
+#define poll(a, b, c) WSAPoll((a), (b), (c))
+#define closesocket(x) closesocket(x)
+typedef int socklen_t;
 
 #ifndef SO_EXCLUSIVEADDRUSE
 #define SO_EXCLUSIVEADDRUSE ((int) (~SO_REUSEADDR))
@@ -106,13 +126,15 @@ typedef int socklen_t;
 #define MG_SOCK_RESET(errcode) \
   (((errcode) < 0) && (WSAGetLastError() == WSAECONNRESET))
 
+#endif  // MG_ENABLE_WINSOCK
+
 #define realpath(a, b) _fullpath((b), (a), MG_PATH_MAX)
-#define sleep(x) Sleep((x) * 1000)
+#define sleep(x) Sleep((x) *1000)
 #define mkdir(a, b) _mkdir(a)
 #define timegm(x) _mkgmtime(x)
 
 #ifndef S_ISDIR
-#define S_ISDIR(x) (((x) & _S_IFMT) == _S_IFDIR)
+#define S_ISDIR(x) (((x) &_S_IFMT) == _S_IFDIR)
 #endif
 
 #ifndef MG_ENABLE_DIRLIST

--- a/src/config.h
+++ b/src/config.h
@@ -56,7 +56,8 @@
 #define MG_ENABLE_MD5 1
 #endif
 
-// Set MG_ENABLE_WINSOCK=0 for Win32 builds with external IP stack (like LWIP)
+// Set MG_ENABLE_WINSOCK=0 for Win32 builds with other external IP stack not
+// mentioned in arch_win32.h
 #ifndef MG_ENABLE_WINSOCK
 #define MG_ENABLE_WINSOCK 1
 #endif

